algorithms/linfa-clustering/examples/dbscan.rs

```diff
@@ -28,7 +28,9 @@ fn main() {
     );
 
     // Infer an optimal set of centroids based on the training data distribution
-    let cluster_memberships = Dbscan::params(min_points).tolerance(1.).predict(dataset);
+    let cluster_memberships: Result<_, _> =
```

Comment:
mhh this is probably caused by the double blanket trait :/ perhaps two separate implementations for the algorithm traits is not ideal and we could circumenvent the problem doing the following:
```rust
pub trait ParameterSet {
    /// Error type resulting from failed hyperparameter checking
    type Error: Error;

    /// Checks the hyperparameters and returns a reference to the checked hyperparameters if
    /// successful
    fn check_ref(&self) -> Result<&Self, Self::Error>;

    /// Checks the hyperparameters and returns the checked hyperparameters if successful
    fn check(self) -> Result<Self, Self::Error>;
}


impl<F: Float> ParameterSet for AppxDbscanParams<F> {
    type Error = AppxDbscanParamsError;

    fn check_ref(&self) -> Result<&Self, Self::Error> {
        if self.is_checked {
            return Ok(&self);
        }

        if self.0.min_points <= 1 {
            Err(AppxDbscanParamsError::MinPoints)
        } else if self.0.tolerance <= F::zero() {
            Err(AppxDbscanParamsError::Tolerance)
        } else if self.0.slack <= F::zero() {
            Err(AppxDbscanParamsError::Slack)
        } else {
            Ok(&self)
        }
    }

    fn check(self) -> Result<Self::Checked, Self::Error> {
        self.check_ref()?;
        self.is_checked = true;
        Ok(self)
    }
}
```
and just constraint the type parameter for `Transform`, `Fit`, `Predict`, `FitIncremental`. This simplifies the trait definitions and general structure a lot by just imposing a single comparison performance regression.

Reply:
I tried your suggested change and still got the same type inference error. It turns out the error isn't caused by the extra impls in `hyperparams.rs`, but the `Predict` impls I added in `impl_dataset.rs`. Those impls are there so that types with fallible `PredictRef` impls also have fallible `Predict` impls. However, this results in `Predict` impls with the same input type but two different output types (fallible and non-fallible), which confuses type inference. Removing those `Predict` impls means that fallible `PredictRef` impls no longer result in correct `Predict` impls (the output type is `Dataset<Result<...>>` instead of `Result<Dataset<...>>`). For this reason, I believe we should only implement `PredictRef` for infallible operations (this is already the case in master). As such, parameter types should not implement `PredictRef` at all, and opt for `Fit` and `Transformer` instead. The only implementers of `PredictRef` should be models, which do not need verification. This also clarifies the relationships between parameters, models, and Linfa traits: parameters `fit` models, parameters also `transform` into results, models `predict` results.

Reply:
ah I see! I'm still thinking about the transform, predict distinction in our implementations. I'm of the opinion that predict is appropriate for most implementations, except where the dataset is actually transformed into a new embedding for PCA or PLS. Perhaps we just can return an error from `predict_ref` and unwrap it by default in `predict`?

Reply:
From what I understand most user code call `predict` anyways, so unwrapping in `predict` just leads to more panics. If a `predict_ref` impl returns a Result then only 2 of the 4 blanket `predict` impls are applied, due to the trait bounds on the return type. Specifically, `predict` will not return `Dataset`s if `predict_ref` returns a result. If we're fine with that we could have some fallible `predict_ref` impls.